### PR TITLE
Ported LowPass and GravityCompensation filters iirob_filters repository

### PR DESCRIPTION
**Pull request by destogl**
- using the dynamic parametrization as implemented in existing gravity_compensation filter
- added testcases for loading parameters

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -21,6 +21,7 @@ endif()
 find_package(ament_cmake REQUIRED)
 find_package(control_msgs REQUIRED)
 find_package(rclcpp REQUIRED)
+find_package(rclcpp_lifecycle REQUIRED)
```
Comment (destogl) [inline CMakeLists.txt:24]: Why do we need this?

```suggestion
```

File: CMakeLists.txt [hunk 2]
```diff
@@ -36,6 +37,7 @@ target_compile_definitions(${PROJECT_NAME} PRIVATE "CONTROL_TOOLBOX_BUILDING_LIB
 ament_target_dependencies(${PROJECT_NAME}
     control_msgs
     rclcpp
+    rclcpp_lifecycle
```
Comment (destogl) [inline CMakeLists.txt:40]:
```suggestion
```

File: include/control_filters/low_pass_filter.hpp [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+    virtual bool update(const T& data_in, T& data_out);
+
+    /** \brief Get most recent parameters */
+    bool updateParameters();
+
+private:
+    /** \brief Dynamic parameter callback activated when parameters change */
+    void parameterCallback();
+    rclcpp::Node::SharedPtr node_;
+    rclcpp::Logger logger_;
```
Comment (destogl) [inline include/control_filters/low_pass_filter.hpp:50]: Do we need this? Isn't there access to logging interface available through `FilterBase`?

File: include/control_filters/low_pass_filter.hpp [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+template <typename T>
+bool LowPassFilter<T>::configure()
+{
+    configure("");
+    return true;
+}
+
+template <typename T>
+bool LowPassFilter<T>::configure(const std::string& ns)
```
Comment (destogl) [inline include/control_filters/low_pass_filter.hpp:90]: this should be merged with `configure()`; as per comment above.

File: include/control_filters/low_pass_filter.hpp [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+    old_value = data_in;
+
+    return true;
+}
+
+template <typename T>
+bool LowPassFilter<T>::updateParameters()
+{
+    
+    if (!filters::FilterBase<T>::getParam(parameter_namespace_ + "sampling_frequency", sampling_frequency_)) {
```
Comment (destogl) [inline include/control_filters/low_pass_filter.hpp:187]: is this `parameter_namespace_` necessary? Shouldn't this work git `getParam` automatically? How is this done in the filters in `ros2/filters` package?

File: test/control_filters/test_low_pass_filter.cpp [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "test_low_pass_filter.hpp"
+
+TEST_F(LowPassFilterTest, TestLowPassFilter)
```
Comment (destogl) [inline test/control_filters/test_low_pass_filter.cpp:17]: Excellent test. Please also add a `load_low_pass_filter_test.cpp` as we have for the controllers. This tests if the plugin is loadable. The best is to add a new file for this.

For an example see [here](https://github.com/ros-controls/ros2_controllers/blob/master/forward_command_controller/test/test_load_forward_command_controller.cpp)

**Review by destogl — COMMENTED** [review]
there are few simplifications to do.

File: control_filters.xml [hunk 7]
```diff
@@ -1,13 +1,22 @@
...
+    <library path="gravity_compensation">
+        <class name="control_filters/GravityCompensatorWrench"
+            type="control_filters::GravityCompensator<geometry_msgs::msg::WrenchStamped>"
+            base_class_type="filters::FilterBase<geometry_msgs::msg::WrenchStamped>">
+            <description>
+                This is a gravity compensation filter working with geometry_msgs::WrenchStamped.
+                It subtracts the influence of a force caused by the gravitational force from force
+                measurements.
+            </description>
+        </class>
```
[the remark this answers is not in the document]
Comment (dzumkeller) [reply control_filters.xml:11]: indentation on my IDE was set to 4 spaces instead of two as the existing code had...

**Review by dzumkeller — APPROVED** [review]

File: src/control_filters/gravity_compensation.cpp [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+  temp_vector_in.vector = data_in.wrench.torque;
+  tf2::doTransform(temp_vector_in, temp_torque_transformed, transform_);
+
+  // Transform CoG Vector
+  geometry_msgs::msg::Vector3Stamped cog_transformed;
+  tf2::doTransform(parameters_->cog_, cog_transformed, transform_cog_);
+
+  // Compensate for gravity force
+  temp_force_transformed.vector.z += parameters_->force_z_;
```
Comment (destogl) [inline src/control_filters/gravity_compensation.cpp:63]: We should always subtract the parameter value independently if it is positive or negative.

File: include/control_filters/gravity_compensation.hpp [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+    }
+
+    return ret;
+  }
+
+  void update_storage() override
+  {
+    world_frame_ = string_parameters_[0].second;
+    RCUTILS_LOG_INFO_NAMED(logger_name_.c_str(), "World frame: %s", world_frame_.c_str());
+    sensor_frame_ = string_parameters_[0].second;
```
Comment (guihomework) [inline include/control_filters/gravity_compensation.hpp:71]:
```suggestion
    sensor_frame_ = string_parameters_[1].second;
```

File: include/control_filters/gravity_compensation.hpp [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+    RCUTILS_LOG_INFO_NAMED(logger_name_.c_str(), "Sensor frame: %s", sensor_frame_.c_str());
+    force_frame_ = string_parameters_[0].second;
```
Comment (guihomework) [inline include/control_filters/gravity_compensation.hpp:73]:
```suggestion
    force_frame_ = string_parameters_[2].second;
```

**Comment by destogl** [comment]
closed in favor of #152 and #153

**Pull request closed without merging**